### PR TITLE
Battle Style Fixes

### PR DESCRIPTION
**Pull request by luongthomas**
🤞

### CODE REVIEW

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy [hunk 1]
```diff
@@ -998,13 +998,18 @@ public enum BattleStyles implements LogicCardInfo {
...
           attackRequirement {}
           onAttack {
-            // TODO
             damage 10
-            def tar = opp.deck.subList(0,6).filterByType(ITEM)
 
-            my.hand.filterByType(ITEM)
-            if(tar){
-              tar.select(max:tar.size(),"Choose the item to discard").discard()
+            def ballCards = my.hand.filterByType(ITEM).filterByNameLike("Ball")
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen8/BattleStyles.groovy:1003]: Make sure that this doesn't match Bursting *Ball*oon, or *Ball*oon Berry. I think most relevant cards end with "Ball"

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy [hunk 2]
```diff
@@ -2123,8 +2132,8 @@ public enum BattleStyles implements LogicCardInfo {
           energyCost FIGHTING, FIGHTING, COLORLESS
           attackRequirement {}
           onAttack {
-            discardAllSelfEnergy(null)
             multiDamage(opp.all, 2, 120)
+            discardAllSelfEnergy(null)
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen8/BattleStyles.groovy:2136]: This has to be in an afterDamage closure. Everything not in an afterDamage closure happens before damage.

**Review by flagrama — CHANGES_REQUESTED** [review]

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy (hunk 2 above)
Comment (luongthomas) [reply src/tcgwars/logic/impl/gen8/BattleStyles.groovy:2136]: Good point, I'll wrap it in an afterDamage closure

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy (hunk 1 above)
Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/BattleStyles.groovy]: Relevant rulings: https://compendium.pokegym.net/category/5-trainers/apricorn-maker/

Comment (luongthomas) [reply src/tcgwars/logic/impl/gen8/BattleStyles.groovy]: Updated the code to check for cards that end in "Ball"

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy (hunk 2 above)
Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/BattleStyles.groovy]: Tegull mentioned some other cards that need this treatment too.

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy [hunk 3]
```diff
@@ -1995,7 +2001,9 @@ public enum BattleStyles implements LogicCardInfo {
                 damage 60
               }
             }
-            top.discard()
+            afterDamage {
+              top.discard()
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen8/BattleStyles.groovy:2005]: You might want to clarify this one. Seems impossible to discard after damage when discarding calculates the damage.

File: src/tcgwars/logic/impl/gen8/BattleStyles.groovy [hunk 4]
```diff
@@ -2177,7 +2185,9 @@ public enum BattleStyles implements LogicCardInfo {
           attackRequirement {}
           onAttack {
             damage 50
-            discardSelfEnergy(C)
+            afterDamage {
+              discardSelfEnergy(C)
```
Comment (flagrama) [inline src/tcgwars/logic/impl/gen8/BattleStyles.groovy:2189]: This will not select energy before damage.

Comment (flagrama) [reply src/tcgwars/logic/impl/gen8/BattleStyles.groovy:2189]: `discardSelfEnergyAfterDamage`

**Pull request merged**